### PR TITLE
#167484612 Super admin can create users

### PR DESCRIPTION
**Pull request by malaba6**
### What does this PR do?
- Creates the endpoint where the super-admin can create a user(Normal users and admins)
- Creates the endpoint where the super admin can change a user Role
- Creates the endpoint where admins can delete a user
- Creates the endpoint where the admins can update user profiles

### Description of Task to be completed?
- In the userController.js file, create a function ‘createUser’ for the Super admin to create a user.            This endpoint (“POST /api/v1/users/createUser”) is protected. User details will include:
First name
Last Name
User name
Email
Password
Confirm Password
Role 
- Edit the  FetchProfiles for the Super Admin to fetch all users (Normal users and Admins). 
- In the userController, create function ‘deleteUser’ where the super admin can delete a user. 
- In the middleware folder, edit the middleware ‘editProfile.js’ to allow the admin and super admin to edit a normal user’s profile. 
- In the userController, create a function for the admin to change a user’s access level. 
- Create tests for these endpoints.
- Add these endpoints to the swagger documentation.

Each role corresponds to a level certain level of authorization for the user in the system. Here are the details (permissions) related to each user role:
- User Role: 
    - User can create their own account 
    - User can can edit their own profile 
    - User can create an article 
    - User can delete their own articles 
    - User can view a specific article 
    - User can view other profiles 
    - Users can follow one another 
    - User can like an article
    - User can comment on an article
    - User can report an article as offensive

- Admin Role:
    - Admin can do all that the user can do 
    - Admin can edit a user’s profile
    - Admin can delete a user
    - Admin can delete an article

- Super-admin Role:
    - Super-admin can Create users (admins and normal users)
    - Super-admin can update a user profile
    - Super admin can change user role
    - Super-admin can see user profiles
    - Super-admin can delete a user
    - Super-admin can delete an article

### How should this be manually tested?
In Postman or Swagger or any other tool of your choice visit these enpoints:
*POST*`/api/v1/users` to create a user
*+PATCH*`/api/v1/users/:username` to update User role
*DELETE*`/api/v1/users/:users` to delete a user
*PUT*`/api/v1/profiles/:username` to update a user profile

### Any background context you want to provide:
To access all these endpoints, The user must be authenticated. And use the token generated for Loggin or Registration as `Authorisation` in the 'header'.

### What are the relevant pivotal tracker stories?
[Super admin can create users(normal users and admins)](https://www.pivotaltracker.com/story/show/#167484612)
[Role based access control](https://www.pivotaltracker.com/story/show/#167484613)

### Screenshots? 
*In Postman*
Screen Shot 2019-08-21 at 11 08 52

*In swagger*
Screen Shot 2019-08-21 at 11 07 20
Screen Shot 2019-08-21 at 11 13 37

### CODE REVIEW

**Review by Jkadhuwa — COMMENTED** [review]
- Looks good but In the documentation, you are using FETCH. I think you meant PATCH.

**Review by raymond42 — CHANGES_REQUESTED** [review]
change the title of the PR to super-admin, not super user. And also in the description for updating a user method is patch not fetch.

**Comment by malaba6** [comment]
> change the title of the PR to super-admin, not super user. And also in the description for updating a user method is patch not fetch.

Done. Thanx!

**Comment by malaba6** [comment]
> * Looks good but In the documentation, you are using FETCH. I think you meant PATCH.

Right! That's what I meant. Thanx, I have changed that

File: package.json [hunk 1]
```diff
@@ -40,6 +45,7 @@
     "mongoose": "^5.2.2",
     "mongoose-unique-validator": "^2.0.1",
     "morgan": "^1.9.1",
+    "npm-run-all": "^4.1.5",
```
Comment (harerakalex) [inline package.json]: This should be dev dependency.

File: package.json [hunk 2]
```diff
@@ -8,11 +8,16 @@
...
+    "mocha-test": "nyc --reporter=text mocha  --require @babel/register --require @babel/polyfill  ./src/tests/*.js --timeout 30000 --exit",
+    "test": "cross-env NODE_ENV=test npm-run-all db:ready mocha-test",
     "down": "cross-env NODE_ENV=test node_modules/.bin/sequelize db:migrate:undo",
+    "db:ready": "npm-run-all down migrate seed",
     "migrate": "cross-env NODE_ENV=test node_modules/.bin/sequelize db:migrate",
-    "seed": "node_modules/.bin/sequelize db:seed:all"
+    "seed": "node_modules/.bin/sequelize db:seed:all",
+    "unseed:all": "node_modules/.bin/sequelize db:seed:undo:all",
+    "undo:seed": "node_modules/.bin/sequelize db:seed:undo:all",
+    "dropdb": "node_modules/.bin/sequelize db:drop",
```
Comment (harerakalex) [inline package.json]: It seems that is not being used. So remove those two script
`"dropdb": "node_modules/.bin/sequelize db:drop",`
`"createdb": "node_modules/.bin/sequelize db:create authors_haven_test"`

**Review by harerakalex — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (malaba6) [reply package.json]: Right! Done!

File: package.json (hunk 1 above)
Comment (malaba6) [reply package.json]: Thanx. done!

File: src/controllers/userController.js [hunk 3]
```diff
@@ -139,6 +154,85 @@ class Authentication {
...
+    try {
+      const name = req.params.username;
+      const userRole = req.userData.role;
+      const user = await User.findOne({ where: { userName: name } });
+      if (!user) {
+        return res.status(404).json({
+          error: `User with username ${name} not found`
+        });
+      }
+      if (userRole === 'user' || ((userRole !== 'admin' || userRole !== 'super-admin')
```
Comment (Jkadhuwa) [inline src/controllers/userController.js]: I think you can use shorthand here.
if (userRole === 'user' || ((userRole !== ('admin' || 'super-admin'))............

File: src/routes/routes.js [hunk 4]
```diff
@@ -11,13 +11,15 @@ import facebookRequest from '../middleware/facebook';
...
 import canEditProfile from '../middleware/editProfile';
+import adminPermissions from '../middleware/adminPermissions';
 import Article from '../controllers/articleController';
 
 const router = express.Router();
 
 const connectMulti = connectmultiparty();
 
 router.post('/api/v1/users', Validation.signupValidation, UserAuth.signup);
+router.post('/api/v1/users/signup', verifyToken, adminPermissions, Validation.signupValidation, UserAuth.signup);
```
Comment (mkiterian) [inline src/routes/routes.js]: its a confusing that the admin endpoint is `/users/signup` and the user registration endpoint is `/users`. I'd recommend we swap the two

**Review by mkiterian — CHANGES_REQUESTED** [review]

File: src/routes/routes.js (hunk 4 above)
Comment (malaba6) [reply src/routes/routes.js]: You are right. Done!

File: src/controllers/userController.js (hunk 3 above)
Comment (malaba6) [reply src/controllers/userController.js]: Yeah! fixed! Thanx...

**Review by harerakalex — APPROVED** [review]

**Pull request merged**